### PR TITLE
TLS bootstrap using tnc

### PR DESCRIPTION
**Pull request by abhinavdahiya**
https://github.com/coreos-inc/tectonic-operators/pull/328

### CODE REVIEW

File: modules/bootkube/assets.tf [hunk 1]
```diff
@@ -25,21 +25,19 @@ resource "template_dir" "bootkube" {
     cluster_cidr          = "${var.cluster_cidr}"
     tectonic_networking   = "${var.tectonic_networking}"
 
-    root_ca_cert                   = "${base64encode(var.root_ca_cert_pem)}"
-    aggregator_ca_cert             = "${base64encode(var.aggregator_ca_cert_pem)}"
-    kube_ca_cert                   = "${base64encode(var.kube_ca_cert_pem)}"
-    kube_ca_key                    = "${base64encode(var.kube_ca_key_pem)}"
-    kubelet_bootstrap_token_id     = "${random_string.kubelet_bootstrap_token_id.result}"
```
Comment (enxebre) [inline modules/bootkube/assets.tf]: The creation of ` resource "random_string" "kubelet_bootstrap_token_id"` and `resource "random_string" "kubelet_bootstrap_token_secret"` at the top of this file can be removed now

File: modules/bootkube/resources/kubeconfig-kubelet [hunk 2]
```diff
@@ -8,8 +8,9 @@ clusters:
...
 - name: kubelet
   user:
-    token: ${kubelet_bootstrap_token_id}.${kubelet_bootstrap_token_secret}
+    client-certificate-data: ${client_cert}
+    client-key-data: ${client_key}
 contexts:
 - context:
     cluster: ${cluster_name}
-    user: kubelet
+    user: kubelet
```
Comment (enxebre) [inline modules/bootkube/resources/kubeconfig-kubelet]: nit: missing EOF new line here

File: modules/tls/kube/kubelet.tf [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  ca_private_key_pem = "${var.kube_ca_key_pem}"
+  ca_cert_pem        = "${var.kube_ca_cert_pem}"
+
+  # want bootstrap node to rotate certificate as soon as possible for the first time
+  validity_period_hours = "1"
+
+  allowed_uses = [
+    "key_encipherment",
+    "digital_signature",
+    "server_auth",
```
Comment (enxebre) [inline modules/tls/kube/kubelet.tf]: does this actually ever get used as sever_auth?

File: steps/assets/ignition-bootstrap.tf [hunk 4]
```diff
@@ -62,6 +63,16 @@ data "ignition_config" "bootstrap" {
    ))}"]
 }
 
+data "ignition_file" "bootstrap_kubeconfig" {
+  filesystem = "root"
+  path       = "/etc/kubernetes/kubeconfig"
+  mode       = 0755
```
Comment (enxebre) [inline steps/assets/ignition-bootstrap.tf]: do we need exec perms here? shouldn't it be 0644 or even 0440, not sure we need it to be writable for some use case but pretty sure we can remove the exec perms

File: installer/pkg/config-generator/generator.go [hunk 5]
```diff
@@ -182,13 +182,12 @@ func (c ConfigGenerator) tncoConfig() (*tnco.OperatorConfig, error) {
...
 	}
 
-	tncoConfig.ControllerConfig.KubeconfigFetchCmd = "" // TODO(yifan): Get kubeconfigFetchCmd.
 	tncoConfig.ControllerConfig.ClusterDNSIP = cidrhost
 	tncoConfig.ControllerConfig.CloudProvider = strings.ToLower(c.Cluster.Platform)
 	tncoConfig.ControllerConfig.CloudProviderConfig = "" // TODO(yifan): Get CloudProviderConfig.
 	tncoConfig.ControllerConfig.ClusterName = c.Cluster.Name
 	tncoConfig.ControllerConfig.BaseDomain = c.Cluster.BaseDomain
-	tncoConfig.ControllerConfig.EtcdInitialCluster = c.etcdInitialCluster()
+	tncoConfig.ControllerConfig.EtcdInitialCount = c.Cluster.NodeCount(c.Cluster.Etcd.NodePools)
```
Comment (enxebre) [inline installer/pkg/config-generator/generator.go]: This requires to update the golang unit tests for config-generator

File: modules/bootkube/assets.tf (hunk 1 above)
Comment (abhinavdahiya) [reply modules/bootkube/assets.tf]: :+1:

File: modules/bootkube/resources/kubeconfig-kubelet (hunk 2 above)
Comment (abhinavdahiya) [reply modules/bootkube/resources/kubeconfig-kubelet]: :+1:

File: modules/tls/kube/kubelet.tf (hunk 3 above)
Comment (abhinavdahiya) [reply modules/tls/kube/kubelet.tf]: no actually, fixed :+1:

File: steps/assets/ignition-bootstrap.tf (hunk 4 above)
Comment (abhinavdahiya) [reply steps/assets/ignition-bootstrap.tf]: changed to 0644 :+1:

File: installer/pkg/config-generator/generator.go (hunk 5 above)
Comment (abhinavdahiya) [reply installer/pkg/config-generator/generator.go]: :+1:

**Comment by abhinavdahiya** [comment]
retest this please. aws only

**Comment by abhinavdahiya** [comment]
retest this please.

**Comment by abhinavdahiya** [comment]
retest this please.

**Comment by trawler** [comment]
@abhinavdahiya both this PR and #3173 both touch on the same area as #3162.
Let's sync up on the priorities

**Comment by enxebre** [comment]
@trawler you probably want an update on how the TNC etcd tls bootstrap

**Comment by abhinavdahiya** [comment]
@yifan-gu @enxebre the tests are green.

File: modules/bootkube/resources/kubeconfig-kubelet [hunk 6]
```diff
@@ -8,8 +8,9 @@ clusters:
 users:
 - name: kubelet
   user:
-    token: ${kubelet_bootstrap_token_id}.${kubelet_bootstrap_token_secret}
+    client-certificate-data: ${client_cert}
```
[the remark this answers is not in the document]
Comment (enxebre) [reply modules/bootkube/resources/kubeconfig-kubelet]: @abhinavdahiya @yifan-gu could you help me to understand this? I see the long running TNC serves the SA token which make sense https://github.com/coreos-inc/tectonic-operators/pull/328/files#diff-5e6686fc064b3b88c5ff3ca4bfbfec84R514 but how is this one restricted

File: modules/bootkube/resources/bootkube.sh [hunk 7]
```diff
@@ -19,8 +19,47 @@ set -e
     --render-output=/assets/tnc-bootstrap
```
Comment (enxebre) [inline modules/bootkube/resources/bootkube.sh]: See discussion https://github.com/coreos/tectonic-installer/pull/3179#issuecomment-382304954

**Comment by enxebre** [comment]
@abhinavdahiya @yifan-gu thanks for addressing the comments, it'd be easier to review/merge by keeping separate PRs for kubeconfig and etcd cert signing. Over all is looking good to me other than https://github.com/coreos/tectonic-installer/pull/3179#issuecomment-382304954 @trawler PTAL

**Comment by enxebre** [comment]
@abhinavdahiya since https://github.com/coreos/tectonic-installer/pull/3183 got it, this needs rebase now, then we can merge

**Comment by aaronlevy** [comment]
Looks like this should be good to go. We're going to merge this from our end so it doesn't get stuck in rebase loop again - if there's any issue we can back it out later.

**Pull request merged**